File: DotNetMessaging/backend/DotNetMessaging.API/Repositories/MessageRepository.cs
using MongoDB.Driver;
using DotNetMessaging.API.Data;
using DotNetMessaging.API.Models;

namespace DotNetMessaging.API.Repositories;

public class MessageRepository : Repository<Message>, IMessageRepository
{
    public MessageRepository(MongoDbContext context) : base(context, "messages")
    {
        // Create indexes
        var chatIndex = Builders<Message>.IndexKeys.Ascending(m => m.ChatId);
        var groupIndex = Builders<Message>.IndexKeys.Ascending(m => m.GroupId);
        var createdAtIndex = Builders<Message>.IndexKeys.Descending(m => m.CreatedAt);
        
        _collection.Indexes.CreateOne(new CreateIndexModel<Message>(chatIndex));
        _collection.Indexes.CreateOne(new CreateIndexModel<Message>(groupIndex));
        _collection.Indexes.CreateOne(new CreateIndexModel<Message>(createdAtIndex));
    }

    public async Task<IEnumerable<Message>> GetChatMessagesAsync(string chatId, int skip = 0, int take = 50)
    {
        var filter = Builders<Message>.Filter.And(
            Builders<Message>.Filter.Eq(m => m.ChatId, chatId),
            Builders<Message>.Filter.Eq(m => m.IsDeleted, false)
        );
        
        var sort = Builders<Message>.Sort.Descending(m => m.CreatedAt);
        
        return await _collection.Find(filter)
            .Sort(sort)
            .Skip(skip)
            .Limit(take)
            .ToListAsync();
    }

    public async Task<IEnumerable<Message>> GetGroupMessagesAsync(string groupId, int skip = 0, int take = 50)
    {
        var filter = Builders<Message>.Filter.And(
            Builders<Message>.Filter.Eq(m => m.GroupId, groupId),
            Builders<Message>.Filter.Eq(m => m.IsDeleted, false)
        );
        
        var sort = Builders<Message>.Sort.Descending(m => m.CreatedAt);
        
        return await _collection.Find(filter)
            .Sort(sort)
            .Skip(skip)
            .Limit(take)
            .ToListAsync();
    }

    public async Task<long> GetUnreadCountAsync(string chatId, string userId, HashSet<string>? readMessageIds = null)
    {
        var filter = Builders<Message>.Filter.And(
            Builders<Message>.Filter.Eq(m => m.ChatId, chatId),
            Builders<Message>.Filter.Ne(m => m.SenderId, userId),
            Builders<Message>.Filter.Eq(m => m.IsDeleted, false)
        );

        if (readMessageIds != null && readMessageIds.Any())
        {
            filter = Builders<Message>.Filter.And(
                filter,
                Builders<Message>.Filter.Nin(m => m.Id, readMessageIds)
            );
        }
        
        return await _collection.CountDocumentsAsync(filter);
    }

    public async Task<long> GetGroupUnreadCountAsync(string groupId, string userId, HashSet<string>? readMessageIds = null)
    {
        var filter = Builders<Message>.Filter.And(
            Builders<Message>.Filter.Eq(m => m.GroupId, groupId),
            Builders<Message>.Filter.Ne(m => m.SenderId, userId),
            Builders<Message>.Filter.Eq(m => m.IsDeleted, false)
        );

        if (readMessageIds != null && readMessageIds.Any())
        {
            filter = Builders<Message>.Filter.And(
                filter,
                Builders<Message>.Filter.Nin(m => m.Id, readMessageIds)
            );
        }
        
        return await _collection.CountDocumentsAsync(filter);
    }
}

File: DotNetMessaging/SpringBootMessaging/src/main/java/com/messaging/api/repositories/MessageRepository.java
package com.messaging.api.repositories;

import com.messaging.api.models.Message;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.data.mongodb.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface MessageRepository extends MongoRepository<Message, String> {
    @Query("{ chatId: ?0, isDeleted: false }")
    Page<Message> findByChatIdOrderByCreatedAtDesc(String chatId, Pageable pageable);
    
    @Query("{ groupId: ?0, isDeleted: false }")
    Page<Message> findByGroupIdOrderByCreatedAtDesc(String groupId, Pageable pageable);
    
    @Query("{ chatId: ?0, senderId: { $ne: ?1 }, isDeleted: false }")
    long countUnreadByChatIdAndUserId(String chatId, String userId);
    
    @Query("{ groupId: ?0, senderId: { $ne: ?1 }, isDeleted: false }")
    long countUnreadByGroupIdAndUserId(String groupId, String userId);
    
    List<Message> findByChatIdAndSenderIdNotAndIsDeletedFalse(String chatId, String userId);
    
    List<Message> findByGroupIdAndSenderIdNotAndIsDeletedFalse(String groupId, String userId);
}

